exchanges/ticker/ticker.go

```diff
@@ -1,193 +1,247 @@
+	AssetType    asset.Item    `json:"assetType"`
+	LastUpdated  time.Time
 }
 
 // Ticker struct holds the ticker information for a currency pair and type
 type Ticker struct {
-	Price        map[string]map[string]map[string]Price
-	ExchangeName string
+	Price
+	Main  uuid.UUID
+	Assoc []uuid.UUID
 }
 
-// PriceToString returns the string version of a stored price field
-func (t *Ticker) PriceToString(p currency.Pair, priceType string, tickerType asset.Item) string {
-	priceType = strings.ToLower(priceType)
-
-	switch priceType {
-	case "last":
-		return strconv.FormatFloat(t.Price[p.Base.Upper().String()][p.Quote.Upper().String()][tickerType.String()].Last, 'f', -1, 64)
-	case "high":
-		return strconv.FormatFloat(t.Price[p.Base.Upper().String()][p.Quote.Upper().String()][tickerType.String()].High, 'f', -1, 64)
-	case "low":
-		return strconv.FormatFloat(t.Price[p.Base.Upper().String()][p.Quote.Upper().String()][tickerType.String()].Low, 'f', -1, 64)
-	case "bid":
-		return strconv.FormatFloat(t.Price[p.Base.Upper().String()][p.Quote.Upper().String()][tickerType.String()].Bid, 'f', -1, 64)
-	case "ask":
-		return strconv.FormatFloat(t.Price[p.Base.Upper().String()][p.Quote.Upper().String()][tickerType.String()].Ask, 'f', -1, 64)
-	case "volume":
-		return strconv.FormatFloat(t.Price[p.Base.Upper().String()][p.Quote.Upper().String()][tickerType.String()].Volume, 'f', -1, 64)
-	case "ath":
-		return strconv.FormatFloat(t.Price[p.Base.Upper().String()][p.Quote.Upper().String()][tickerType.String()].PriceATH, 'f', -1, 64)
+// Update updates ticker price
+func (s *Service) Update(p *Price) error {
+	var ids []uuid.UUID
+
+	s.Lock()
+	switch {
+	case s.Tickers[p.ExchangeName] == nil:
+		s.Tickers[p.ExchangeName] = make(map[*currency.Item]map[*currency.Item]map[asset.Item]*Ticker)
+		s.Tickers[p.ExchangeName][p.Pair.Base.Item] = make(map[*currency.Item]map[asset.Item]*Ticker)
+		s.Tickers[p.ExchangeName][p.Pair.Base.Item][p.Pair.Quote.Item] = make(map[asset.Item]*Ticker)
+		err := s.SetNewData(p)
+		if err != nil {
+			s.Unlock()
+			return err
+		}
+
+	case s.Tickers[p.ExchangeName][p.Pair.Base.Item] == nil:
+		s.Tickers[p.ExchangeName][p.Pair.Base.Item] = make(map[*currency.Item]map[asset.Item]*Ticker)
+		s.Tickers[p.ExchangeName][p.Pair.Base.Item][p.Pair.Quote.Item] = make(map[asset.Item]*Ticker)
+		err := s.SetNewData(p)
+		if err != nil {
+			s.Unlock()
+			return err
+		}
+
+	case s.Tickers[p.ExchangeName][p.Pair.Base.Item][p.Pair.Quote.Item] == nil:
+		s.Tickers[p.ExchangeName][p.Pair.Base.Item][p.Pair.Quote.Item] = make(map[asset.Item]*Ticker)
+		err := s.SetNewData(p)
+		if err != nil {
+			s.Unlock()
+			return err
+		}
+
+	case s.Tickers[p.ExchangeName][p.Pair.Base.Item][p.Pair.Quote.Item][p.AssetType] == nil:
+		err := s.SetNewData(p)
+		if err != nil {
+			s.Unlock()
+			return err
+		}
+
 	default:
-		return ""
+		s.Tickers[p.ExchangeName][p.Pair.Base.Item][p.Pair.Quote.Item][p.AssetType].Last = p.Last
```

Comment:
By doing this, we make the exchangename case sensitive. This does not get flagged by `validExchange(exch string) bool ` as it does `strings.Equalfold`
As a result this works:
` .\gctcli.exe gettickerstream BTSE BTC-USD spot`
but this doesn't:
` .\gctcli.exe gettickerstream btse BTC-USD spot`

This does not impact the currency pair.
This does impact asset type
This does impact getorderbookstream too